File: 0994-rotting-oranges/0994-rotting-oranges.cpp
// class Solution {
// public:
//     int row[4] = {1, 0, -1, 0};
//     int col[4] = {0, 1, 0, -1};
//     int orangesRotting(vector<vector<int>>& grid) {
//         int time = 0;
//         int n = grid.size();
//         int m = grid[0].size();
//         queue<pair<int, int>> Q;
//         for(int i = 0; i < n; i++){
//             for(int j = 0; j < m; j++){
//                 if(grid[i][j] == 2){
//                     Q.push({i, j});
//                 }
//             }
//         }
//         Q.push({-1, -1});
//         // return time;
//         while(Q.size() > 1){
//             auto frontElem = Q.front();
//             Q.pop();
//             for(int i  = 0; i < 4; i++){
//                 int x = frontElem.first+row[i];
//                 int y = frontElem.second+col[i];
//                 if(x >= 0 && x < n && y >= 0 && y < m && grid[x][y] == 1){
//                     grid[x][y] = 2;
//                     Q.push({x, y});
//                 }
//             }
//             if (frontElem.first == -1 && frontElem.second == -1){
//                 Q.push({-1, -1});
//                 time++;
//                 continue;
//             }  
//         }
//         // return time;
//         for(int i = 0; i < n; i++){
//             for(int j = 0; j < m; j++){
//                 if(grid[i][j] == 1){
//                     return -1;
//                 }
//             }
//         }
//         return time;
//     }
// };


class Solution {
public:
    int row[4] = {1, 0, -1, 0};
    int col[4] = {0, 1, 0, -1};
    int orangesRotting(vector<vector<int>>& grid) {
        int time = 0;
        int n = grid.size();
        int m = grid[0].size();
        queue<pair<int, int>> Q;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == 2){
                    Q.push({i, j});
                }
            }
        }
        while(!Q.empty()){
            int size = Q.size();
            int can_rot = 0;
            for(int i = 0; i < size; i++){
                auto frontElem = Q.front();
                Q.pop();
                for(int j  = 0; j < 4; j++){
                    int x = frontElem.first+row[j];
                    int y = frontElem.second+col[j];
                    if(x >= 0 && x < n && y >= 0 && y < m && grid[x][y] == 1){
                        grid[x][y] = 2;
                        Q.push({x, y});
                        can_rot = 1;
                    }
                }
            }
            time += can_rot;
        }
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == 1){
                    return -1;
                }
            }
        }
        return time;
    }
};

File: 0994-rotting-oranges/0994-rotting-oranges.py
# from collections import deque

# class Solution:
#     def orangesRotting(self, grid: List[List[int]]) -> int:
#         n, m = len(grid), len(grid[0])
#         directions = [
#             (1, 0),   # down
#             (-1, 0),  # up
#             (0, 1),   # right
#             (0, -1)   # left
#         ]


#         queue = deque()
#         fresh = 0

#         for i in range(n):
#             for j in range(m):
#                 if grid[i][j] == 2:
#                     queue.append((i, j, 0))
#                 elif grid[i][j] == 1:
#                     fresh += 1

#         max_time = 0
#         while queue:
#             x, y, t = queue.popleft()
#             max_time = max(max_time, t)

#             for dx, dy in directions:
#                 nx, ny = x + dx, y + dy
#                 if 0 <= nx < n and 0 <= ny < m and grid[nx][ny] == 1:
#                     grid[nx][ny] = 2
#                     fresh -= 1
#                     queue.append((nx, ny, t + 1))


#         return max_time if fresh == 0 else -1


from collections import deque

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])

        directions = [(1,0), (-1,0), (0,1), (0,-1)]

        queue = deque()
        fresh = 0

        # initialize
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 2:
                    queue.append((i, j))
                elif grid[i][j] == 1:
                    fresh += 1

        if not fresh:
            return 0
        if not queue:
            return -1

        time = 0
        while queue and fresh > 0:
            for _ in range(len(queue)):  # process one level
                x, y = queue.popleft()

                for dx, dy in directions:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < n and 0 <= ny < m and grid[nx][ny] == 1:
                        grid[nx][ny] = 2
                        fresh -= 1
                        queue.append((nx, ny))

            time += 1

        return time if fresh == 0 else -1